smartconfig/configuration.py

```diff
@@ -0,0 +1,173 @@
+from smartconfig.register import _register
+from smartconfig.typehints import _EntryMapping, _FilePath
+
+
+def load_config_file(path: _FilePath) -> None:
+    """
+    Load a YAML-like configuration file and update configuration entries.
+
+    The values set in the YAML-like file will override values already defined in the `ConfigEntry`.
+    For each section, the name of previous sections will be concatenated in order to make the full path of the entry
+    to override.
+
+    See `smartconfig.YamlLikeParser` for more details on the syntax.
+
+    Args:
+        path: The path to the configuration file. Can be a string or an object defining `os.PathLike`.
+
+    Raises:
+        MalformedYamlFile: The configuration file is malformed.
+        FileNotFoundError: The configuration file doesn't exist.
+        IOError: An error occurred when reading the file.
+    """
+    with open(path) as file:
+        patch = YamlLikeParser(file.read(), file.name).parse()
+
+    for path, entries in patch.items():
+        # Patch every already existing entries.
+        if path in _register.configuration_for_module:
+            _register.configuration_for_module[path]._patch_entries(entries)
+
+        # Update the global register for future entries to use.
+        if path not in _register.global_configuration:
+            _register.global_configuration[path] = {}
+        _register.global_configuration[path].update(entries)
+
+
+class _ConfigEntryMeta(type):
+    """
+    Metaclass used to define special ConfigEntry behaviors.
+
+    Note: Using this metaclass outside of the library is currently not supported.
+    """
+
+    def __getattribute__(cls, item: str):
+        """
+        Special attribute lookup function.
+
+        If the attribute name starts with `_`, normal lookup is done, otherwise _configuration_mapping is used.
+
+        Args:
+            item: Attribute to lookup
+
+        Raises:
+            ConfigurationKeyError: The item doesn't exist.
+        """
+        # Use the normal lookup for attribute starting with `_`
+        if item.startswith('_'):
+            return super().__getattribute__(item)
+
+        if item not in cls._configuration_mapping:
+            raise ConfigurationKeyError(f"Entry {cls._path} doesn't define any {item} entry.")
+
+        return cls._configuration_mapping[item]
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any]) -> type:
+        """Add special attribute to the new entry."""
+        configuration_mapping = {key: value for key, value in dict_.items() if not key.startswith('_')}
+        defined_entries = {
+            name
+            for name
+            in chain(
+                dict_.keys(),
+                dict_.get('__annotations__', {}).keys()
+            )
+            if not name.startswith('_')
+        }
+
+        dict_["_configuration_mapping"] = configuration_mapping
+        dict_["_defined_entries"] = defined_entries
```

Comment:
Name mangling indeed doesn't work when injecting variables inside the dictionary, which is a bit of a shame. Ideally, users should not define attributes starting with an underscore since they are resolved using the normal `__dict__` attribute. Maybe we could add some code to forbid that?

Reply:
The name could manually be mangled when stored in the dict. The metaclass knows what the class's name will be so this should be pretty simple.

Reply:
I tried to manually mangle the name, although because the attributes are looked up in `ConfigEntry` and `ConfigEntryMeta`, it will try to use one or the other to mangle the name, so it will use two names to resolve the same name, which just brings a ton of other issues.

Reply:
Can you move everything over to the metaclass? Then the class will just be a simple "wrapper" to avoid using a metaclass directly. Granted, such change means one will have to subclass the metaclass to customise it but I don't see that as a problem.